### PR TITLE
Design: diagnostic attribute error

### LINKED ISSUE #2235: Support contracts based argument checking
**Is your feature request related to a problem? Please describe.**
Recently, we are making [a great effort in improving diagnostics and type checking](https://github.com/sql-machine-learning/sqlflow/labels/diagnostics) to enhance the experience and development efficiency of SQLFlow users. 

**Describe the solution you'd like**
Contract programming is a great idea that encourages defining formal, precise, and verifiable interface specifications for software components (function or a callable object, for example). 
> ... which extends the ordinary definition of abstract data types with preconditions, preconditions, and invariants. These specifications are referred to as "contracts", in accordance with a conceptual metaphor with the conditions and obligations of business contracts.

--- From https://en.wikipedia.org/wiki/Design_by_contract

The term 'contract programming' was coined by Bertrand Meyer in connection with his design of the Eiffel programming language. The idea is widely accepted by many other programming languages with either native support (`D`, `Scala` and many others) or 3rd-party support (`C++`  via `Boost.Contract`, `Python` via `PyContracts`, `Go`, `Java` and almost all popular languages).

In fact, contract-based verification is exactly what `SQLFlow` is doing in the [attribute](https://github.com/sql-machine-learning/sqlflow/tree/develop/pkg/sql/codegen/attribute) pkg. 

Checking the contract requirements for python functions or callables in golang is ugly, unnatural, boring, and sometimes impossible.

We propose developing a python-native contracts solution that is comprehensive,  precise, concise, and elegant.

The solution should:
1. Support checking a user-defined model via:
    - docstring
    - or decorator
    - or both
1. Support checking existing models like `XGBoost` or canned `Estimators`
1. Give precise and readable diagnostic messages to users that violate the contract requirements

Later on this day, I'll push a concise implementation for demonstration and further discussion.

### PR DESCRIPTION
**Pull request by Yancey0623**
Partially fix #2235

### CODE REVIEW

File: doc/design/diag_attribute_error.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+SELECT * FROM iris.test
+TO PREDICT iris.pred.class
+USING my_model;
+```
+
+SQLFlow compiles the above SQL program into an execution plan and runs it.  As the TRAIN statement above, SQLFlow uses `TO TRAIN CLAUSE`  to train a specific model called  `DNNClassifier`, using `WITH CLAUSE` to configure the training arguments. 
+
+Sometimes users may make some configuration mistake on `WITH CLAUSE`, then the job would fault during execution and return some uncertain error message.
+
+This documentation issued a way that adding extra contract in [docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) to contract the arguments, and this can achieve three advantage at least:
```
Comment (typhoonzero) [inline doc/design/diag_attribute_error.md]: Explain what contract is?

File: doc/design/diag_attribute_error.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+An example:
+
+```python 
+class MyDNNClassifier(keras.Model)
+    def __init__(self, n_classes=32, hidden_units=[32, 64]):
+    """
+    Args:
+
+    # isintance(n_classes, int) && n_classes > 1
```
Comment (typhoonzero) [inline doc/design/diag_attribute_error.md:55]: Google Python Style Docstring defines a method of constraining parameter types. It doesn't constraint parameter value range however, but we can extend the format to constrain the value range.

I propose this approach because I feel that @Yancey1989 's proposal shares the same goal as Google Style Docstring.  I am considering how to combine the two ideas.
 
This is a line of code, not some hints. I would recommend we follow the format of: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html, e.g.

```python
class ExampleError(Exception):
    """Exceptions are documented in the same way as classes.

    The __init__ method may be documented in either the class level
    docstring, or as a docstring on the __init__ method itself.

    Either form is acceptable, but the two should not be mixed. Choose one
    convention to document the __init__ method and be consistent with it.

    Note:
        Do not include the `self` parameter in the ``Args`` section.

    Args:
        msg (str, required): Human readable string describing the exception.
        code (:obj:`int`, optional): Error code.
        init_val (int, optional, (0, 1000]): initial value.

    Attributes:
        msg (str): Human readable string describing the exception.
        code (int): Exception error code.

    """
```

Comment (shendiaomo) [reply doc/design/diag_attribute_error.md:55]: > This is a line of code, not some hints. I would recommend we follow the format of: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html, e.g.
> 
> ```python
> class ExampleError(Exception):
>     """Exceptions are documented in the same way as classes.
> 
>     The __init__ method may be documented in either the class level
>     docstring, or as a docstring on the __init__ method itself.
> 
>     Either form is acceptable, but the two should not be mixed. Choose one
>     convention to document the __init__ method and be consistent with it.
> 
>     Note:
>         Do not include the `self` parameter in the ``Args`` section.
> 
>     Args:
>         msg (str, required): Human readable string describing the exception.
>   ```
A critical flaw of this method is that we never know what the message is, hence we'll have no control of the user experience. For example:
```python
Args: 
    n_classes (int, required, [2, )): should be >=20
```
In contrast, #2245 has a solid mathematical theory basis. If we really have to parse requirements in the docstrings. A contributor can write:
##### Snippet 1
```python
"""
    :param: n_classes: (Describe the purpose of `n_classes`)
    :require: Int & Greater(1)   
"""
```
This would guarantee the consistency of what the author wants and what the users think what the author wants.
The snippet `Int & Greater(1)` will be substituted by `contracts` to the human-readable form. Therefore, the end-users will never see `Int & Greater(1)`, but `int that's > 1`.

Moreover, the messages generated by `contracts.py` are actually DNFs in plain english, we can easily parse them back to DNFs. With the help of `contracts.py` (and a simple parser implemented in `split`), a contributor can even write:
##### Snippet 2
```python
"""
    :param: n_classes: (Describe the purpose of `n_classes`)
    :require: int that's >= 1   
"""
```
Frankly speaking, if you believe writing requirements in docstring is a good idea, I think **Snippet 1** is enough.

File: doc/design/diag_attribute_error.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    classification. Must be > 1.
+
+    # isintance(hidden_units, list) && all(isinstance(item, int) for item in hidden_units)
+    hidden_units: Iterable of number hidden units per layer. All layers are
+    fully connected. Ex. `[64, 32]` means first layer has 64 nodes and
+    second one has 32.
+    """
+```
+
+If a user enter some invalide arguments:
```
Comment (shendiaomo) [inline doc/design/diag_attribute_error.md]: invalide -> invalid

File: doc/design/diag_attribute_error.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  model.hidden_units = [128, 64]
+LABEL class
+INTO my_model;
+
+SELECT * FROM iris.test
+TO PREDICT iris.pred.class
+USING my_model;
+```
+
+SQLFlow compiles the above SQL program into an execution plan and runs it.  As the TRAIN statement above, SQLFlow uses `TO TRAIN CLAUSE`  to train a specific model called  `DNNClassifier`, using `WITH CLAUSE` to configure the training arguments.
```
Comment (wangkuiyi) [inline doc/design/diag_attribute_error.md]: SQLFlow compiles each statement in the program into an execution plan and executes them.

Comment (wangkuiyi) [inline doc/design/diag_attribute_error.md]: `TO TRAIN CLAUSE` => `TO TRAIN` clause

Comment (wangkuiyi) [inline doc/design/diag_attribute_error.md]: `WITH` clause

File: doc/design/diag_attribute_error.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+INTO my_model;
+
+SELECT * FROM iris.test
+TO PREDICT iris.pred.class
+USING my_model;
+```
+
+SQLFlow compiles the above SQL program into an execution plan and runs it.  As the TRAIN statement above, SQLFlow uses `TO TRAIN CLAUSE`  to train a specific model called  `DNNClassifier`, using `WITH CLAUSE` to configure the training arguments. 
+
+Sometimes users may make some configuration mistake on `WITH CLAUSE`, then the job would fault during execution and return some uncertain error message.
```
Comment (wangkuiyi) [inline doc/design/diag_attribute_error.md]: The word fault is a noun, not a verb.

File: doc/design/diag_attribute_error.md (hunk 1 above)
Comment (wangkuiyi) [reply doc/design/diag_attribute_error.md]: We don't need the word contract here.  We can say 

The model parameter documentation describes parameters and acceptable values of human reading.  We want to enhance it for the reading by the SQLFlow compiler so to warn about wrongly set parameters.

File: doc/design/diag_attribute_error.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+This documentation issued a way that adding extra contract in [docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) to contract the arguments, and this can achieve three advantage at least:
+
+1. Early testing, we can do early testing before running the job; users can wait less time and cluster save resources.
+2. More accurate diagnostic message.
+3. Using docstring is Keras or Tensorflow native; users don't need to modify the model code.
```
Comment (wangkuiyi) [inline doc/design/diag_attribute_error.md]: What does this mean "Using docstring is Keras or Tensorflow native;"?

File: doc/design/diag_attribute_error.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+## Design
+
+The docstring of a function or a class can include argument documentation and contracts.
```
Comment (wangkuiyi) [inline doc/design/diag_attribute_error.md]: Do you want to say "We want to document the compiler-readable description of model parameters in the docstring of Python function or class that define a model"?

**Review by wangkuiyi — COMMENTED** [review]
Thanks @Yancey1989 for this design. I think it is very elegant.

1. `# isintance(hidden_units, list) && all(isinstance(item, int) for item in hidden_units)` is very readable. Even a naive Python programmer could understand it.

1. Describe the checking rules in docstring is a brilliant idea because comments were used to make human users understand the parameters, now we reuse it to serve the compiler as well.

I have some comments for minor stuff like English writing.

File: doc/design/diag_attribute_error.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      if !contract.check(name, value):
+            # component received value and argument documentation
+              diag_err.append_message(contract.diag_message(name, value))
+    if !diag_err.empty():
+          raise diag_err
+```
+
+## Future
+
+This documentation using natvie Python code to express the contract, [another PR](https://github.com/sql-machine-learning/sqlflow/pull/2245) issued an DSL, will make more discussion in the future.
```
Comment (wangkuiyi) [inline doc/design/diag_attribute_error.md]: Replace the word "issued" with a meaningful word.

File: doc/design/diag_attribute_error.md (hunk 2 above)
Comment (wangkuiyi) [reply doc/design/diag_attribute_error.md:55]: > ```python
> """
>     :param: n_classes: (Describe the purpose of `n_classes`)
>     :require: int that's >= 1   
> """
> ```

Do you @shendiaomo  propose some special format for the "contract" description? I think the format like `:param:` is delicate and fragile.  How if the users forgot to type a colon `:`? Do we need another layer of contract to describe this contract?

I understand @typhoonzero 's point is that we need compiler readable checking rules and human-readable explanation.  For example, `isintance(n_classes, int) && n_classes > 1` is not enough -- we also need `n_class must be an integer and larger than 1`.

It looks to me that @Yancey1989 already put this under consideration -- each parameter has a documentation of two lines:

- A line starting with `#` is the compiler-readable checking rule

- The next line (optional) is the human-readable rule.

I think @Yancey1989 's design is very complete.  Did I miss something? @typhoonzero @shendiaomo

**Review by wangkuiyi — DISMISSED** [review]
Let us do this approach and close https://github.com/sql-machine-learning/sqlflow/pull/2245.

Let me summarize this proposed approach.

- We use docstring to describe checking rules for parameters.
- A docstring contains zero, one, or more lines.
- A line starting with `#` is a check rule in Python.
- The checking algorithm goes over the lines in the docstring.  Encountering a line starting with `#`, the algorithm executes the checking rule.  If the rule returns True, then it goes on looking for the next `#`-started line; otherwise, it looks for the next non-empty line -- if this line is not `#`-started, prints it as the error message; otherwise, prints the checking rule itself as the error message.

The checking rules and user-defined models are all in the same source file, though the checking rules are in docstrings. So, let us DO NOT run the checking rules and the model definition in the SQLFlow server container for security reasons.  Instead, let us run them in step containers.

File: doc/design/diag_attribute_error.md (hunk 6 above)
Comment (Yancey0623) [reply doc/design/diag_attribute_error.md]: I polished this sentence. 
> Model developers do not have to involve dependencies other than Keras or Tensorflow.

File: doc/design/diag_attribute_error.md (hunk 2 above)
Comment (Yancey0623) [reply doc/design/diag_attribute_error.md:55]: From @typhoonzero 's comment
>         init_val (int, optional, (0, 1000]): initial value.
This case is very simple if we want to express more complex rules such as `x should be an int list and each element greater then 1`, we should extend the syntax.

As the discussion on the daily meeting, to design a Python library instead of a native Python code to check the parameters can make the code shorter and simpler, but we should collective many user cases and then writing a new design of this Python library in the future.

**Review by wangkuiyi — APPROVED** [review]

File: doc/design/diag_attribute_error.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+using `WITH` clause to configure the training arguments.
+
+Sometimes users may make some configuration mistake on `WITH` clause,
+then the job would fail during execution and return some uncertain error message.
+
+The model parameter documentation describes parameters and acceptable values of human reading.
+We want to enhance it for the reading by the SQLFlow compiler so to warn about wrongly set parameters, and
+this can active three advantages at least:
+
+1. Early testing, we can do early testing before running the job; users can wait less time and cluster save resources.
```
Comment (weiguoz) [inline doc/design/diag_attribute_error.md:32]: Plus, SQLFlow enhances the ability of code hints for GUI.

Comment (wangkuiyi) [inline doc/design/diag_attribute_error.md:32]: I am afraid that for security reasons, we cannot early check the values because we can only run user code (in either docstring or in Python source files) in the step containers, but not the SQLFlow server container?

Comment (Yancey0623) [reply doc/design/diag_attribute_error.md:32]: Yes, but for the built-in models(Tensorflow Estimator and XGBoost ), I think we can still check them on the SQLFlow server because we can review the code.
